### PR TITLE
QSP10, QSP11 - Add proper locking

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**
> Code health improvements

**What does this PR do? Why is it needed?**
Addressed locking related feedbacks from Audit report. They are qsp10 and qsp11

**Which issues(s) does this PR fix?**
Part of #6327

### CODE REVIEW

**Comment by mcdee** [comment]
Perhaps consider #6326 instead of the proposed changes to the state here, as they cover a wider range of fixes.

**Comment by terencechain** [comment]
> Perhaps consider #6326 instead of the proposed changes to the state here, as they cover a wider range of fixes.

It seems like #6326 will be taking longer to merge and maybe experimental. (e.g. feature flag was requested) I say we merge this first (lower risk) then default to #6326 after more testing there

File: beacon-chain/state/getters.go [hunk 1]
```diff
@@ -345,6 +352,10 @@ func (b *BeaconState) Eth1DataVotes() []*ethpb.Eth1Data {
 	if b.state.Eth1DataVotes == nil {
 		return nil
 	}
+
+	b.lock.RLock()
```
Comment (prestonvanloon) [inline beacon-chain/state/getters.go:356]: This needs to go before we read `b.state.Eth1DataVotes`

File: beacon-chain/state/getters.go [hunk 2]
```diff
@@ -413,13 +428,13 @@ func (b *BeaconState) ValidatorAtIndex(idx uint64) (*ethpb.Validator, error) {
 	if b.state.Validators == nil {
 		return &ethpb.Validator{}, nil
 	}
-	if uint64(len(b.state.Validators)) <= idx {
-		return nil, fmt.Errorf("index %d out of range", idx)
-	}
 
 	b.lock.RLock()
```
Comment (prestonvanloon) [inline beacon-chain/state/getters.go:432]: This needs to go before we read `b.state.Validators`.

File: beacon-chain/state/setters.go [hunk 3]
```diff
@@ -155,11 +155,11 @@ func (b *BeaconState) UpdateStateRootAtIndex(idx uint64, stateRoot [32]byte) err
 	if !b.HasInnerState() {
 		return ErrNilInnerState
 	}
+	b.lock.RLock()
 	if len(b.state.StateRoots) <= int(idx) {
 		return errors.Errorf("invalid index provided %d", idx)
```
Comment (prestonvanloon) [inline beacon-chain/state/setters.go:160]: You need to unlock in this branch now

File: beacon-chain/state/setters.go [hunk 4]
```diff
@@ -174,7 +175,6 @@ func (b *BeaconState) UpdateStateRootAtIndex(idx uint64, stateRoot [32]byte) err
 		ref.MinusRef()
 		b.sharedFieldReferences[stateRoots] = &reference{refs: 1}
 	}
-	b.lock.RUnlock()
```
Comment (prestonvanloon) [inline beacon-chain/state/setters.go:177]: This should stay here

**Pull request merged**